### PR TITLE
[lldb] Fix ThreadList assignment race

### PR DESCRIPTION
**Pull request by labath**
ThreadList uses the Process mutex to guard its state. This means its not possible to safely modify its process member, as the member is required to lock the mutex.

Fortunately for us, we never actually need to change the process member (we always just juggle different kinds of thread lists belonging to the same process).

This patch replaces the process member assignment (which is technically a race even when it assigns the same value) with an assertion.

Since all this means that the class can never change its process member value (and it also must be non-null at all times), I've also changed the member type to a reference.

### CODE REVIEW

**Comment by labath** [comment]
Previous attempt to fix this issue: https://reviews.llvm.org/D158034

**Review by JDevlieghere — APPROVED** [review]
LGTM. 

As a nice to have, maybe add a short explanation to the first mutex explaining that these operations are only allowed between thread lists belonging to the same process. Your current comment makes it sound as if that only applies *if* the processes are the same, but per the assertion, the requirement is stricter than that. 

EDIT: Oh I see you had the comment in the header. That covers my concern, though I'd still add the same thing to the assert to quickly diagnose the issue if someone were to hit it.

File: lldb/source/Target/ThreadList.cpp [hunk 1]
```diff
@@ -736,11 +734,12 @@ void ThreadList::NotifySelectedThreadChanged(lldb::tid_t tid) {
 
 void ThreadList::Update(ThreadList &rhs) {
   if (this != &rhs) {
-    // Lock both mutexes to make sure neither side changes anyone on us while
-    // the assignment occurs
-    std::scoped_lock<std::recursive_mutex, std::recursive_mutex> guard(GetMutex(), rhs.GetMutex());
+    // Same process implies same mutex...
+    assert(&m_process == &rhs.m_process);
+    assert(&GetMutex() == &GetMutex());
```
Comment (augusto2112) [inline lldb/source/Target/ThreadList.cpp:739]: Should this be:

```
assert(&GetMutex() == &rhs.GetMutex());
```

?

File: lldb/include/lldb/Target/ThreadList.h [hunk 2]
```diff
@@ -27,12 +27,13 @@ class ThreadList : public ThreadCollection {
...
 
 public:
-  ThreadList(Process *process);
+  ThreadList(Process &process);
 
   ThreadList(const ThreadList &rhs);
 
   ~ThreadList() override;
 
+  // Precondition: both thread lists must be belong to the same process.
```
Comment (augusto2112) [inline lldb/include/lldb/Target/ThreadList.h:36]: Should these be `///` so they're picked up by doxygen?

**Review by jimingham — APPROVED** [review]
ThreadList differs from ThreadCollection in that it holds a process, and all the threads in the list must belong to that process.  Things like having "ShouldStop" methods on ThreadList would make no sense for a list of threads from random processes.  We don't ever reuse these as processes come and go in a target.  So having the process seem optional or changeable in ThreadList was a mistake (probably mine).
This is more correct.

File: lldb/source/Target/ThreadList.cpp (hunk 1 above)
Comment (labath) [reply lldb/source/Target/ThreadList.cpp:739]: yes 🤦

File: lldb/include/lldb/Target/ThreadList.h (hunk 2 above)
Comment (labath) [reply lldb/include/lldb/Target/ThreadList.h:36]: Works for me

**Comment by labath** [comment]
> LGTM.
> 
> As a nice to have, maybe add a short explanation to the first mutex explaining that these operations are only allowed between thread lists belonging to the same process. Your current comment makes it sound as if that only applies _if_ the processes are the same, but per the assertion, the requirement is stricter than that.
> 
> EDIT: Oh I see you had the comment in the header. That covers my concern, though I'd still add the same thing to the assert to quickly diagnose the issue if someone were to hit it.

Elaborated the assert comment as well.

**Pull request merged**